src/catalog/src/information_schema/predicate.rs

```diff
@@ -0,0 +1,622 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use arrow::array::StringArray;
+use arrow::compute::kernels::comparison;
+use common_query::logical_plan::DfExpr;
+use datafusion::common::ScalarValue;
+use datafusion::logical_expr::expr::Like;
+use datafusion::logical_expr::Operator;
+use datatypes::value::Value;
+use store_api::storage::ScanRequest;
+
+type ColumnName = String;
+/// Predicate to filter `information_schema` tables stream,
+/// we only support these simple predicates currently.
+/// TODO(dennis): supports more predicate types.
+#[derive(Clone, PartialEq, Eq, Debug)]
+enum Predicate {
+    Eq(ColumnName, Value),
+    Like(ColumnName, String, bool),
+    NotEq(ColumnName, Value),
+    InList(ColumnName, Vec<Value>),
+    And(Box<Predicate>, Box<Predicate>),
+    Or(Box<Predicate>, Box<Predicate>),
+    Not(Box<Predicate>),
+}
+
+impl Predicate {
+    /// Evaluate the predicate with the row, returns:
+    /// - `None` when the predicate can't evaluate with the row.
+    /// - `Some(true)` when the predicate is satisfied,
+    /// - `Some(false)` when the predicate is not satisfied,
+    fn eval(&self, row: &[(&str, &Value)]) -> Option<bool> {
+        match self {
+            Predicate::Eq(c, v) => {
+                for (column, value) in row {
+                    if c != column {
+                        continue;
+                    }
+                    return Some(v == *value);
+                }
+            }
+            Predicate::Like(c, pattern, case_insensitive) => {
+                for (column, value) in row {
+                    if c != column {
+                        continue;
+                    }
+
+                    let Value::String(bs) = value else {
+                        continue;
+                    };
+
+                    return like_utf8(bs.as_utf8(), pattern, case_insensitive);
+                }
+            }
+            Predicate::NotEq(c, v) => {
+                for (column, value) in row {
+                    if c != column {
+                        continue;
+                    }
+                    return Some(v != *value);
+                }
+            }
+            Predicate::InList(c, values) => {
+                for (column, value) in row {
+                    if c != column {
+                        continue;
+                    }
+                    return Some(values.iter().any(|v| v == *value));
+                }
+            }
+            Predicate::And(left, right) => {
+                let left = left.eval(row);
+
+                // short-circuit
+                if matches!(left, Some(false)) {
+                    return Some(false);
+                }
+
+                return match (left, right.eval(row)) {
+                    (Some(left), Some(right)) => Some(left && right),
+                    (None, Some(false)) => Some(false),
```

Comment:
"None" means the Predicate can't eval, kind of "undefined behavior". So I think "None AND false" can't be evalated to false. Also for the "OR" Predicate below.

Reply:
This is a `and` expression, if the left or right expression returns false, the final result is false, even though one of them can't be evaluated. So the same with `OR`.

Reply:
`None` means the predicate can't work on this column currently.  And datafusion already ensures that all columns exist in the schema, so we don't have to worry about other cases.